OctoAwesome/OctoAwesome/Components/ComponentList.cs

```diff
@@ -9,202 +8,360 @@
+            component = default;
+            return false;
         }
 
-        /// <summary>
-        /// Removes the Component of the given type.
-        /// </summary>
-        /// <typeparam name="V">The type of the component to remove.</typeparam>
-        /// <returns>A value indicating whether the remove was successful or not.</returns>
-        public bool RemoveComponent<V>() where V : T
+        component = GenericCaster<T, V>.Cast(result[0]);
+        return contains;
+    }
+
+    /// <summary>
+    /// Returns the first Component of the given Type or null
+    /// </summary>
+    /// <typeparam name="V">Component Type</typeparam>
+    /// <returns>Component</returns>
+    public V? Get<V>()
+    {
+        if (componentsByType.TryGetValue(typeof(V), out var result) && result.Count > 0)
+            return GenericCaster<T, V>.Cast(result[0]);
+
+        return default;
+    }
+
+    /// <summary>
+    /// Returns a list of the Component with the given Type or null
+    /// </summary>
+    /// <typeparam name="V">Component Type</typeparam>
+    /// <returns>Component</returns>
+    public List<V>? GetAll<V>()
+    {
+        if (componentsByType.TryGetValue(typeof(V), out var result))
+            return GenericCaster<T, V>.CastList<List<T>, List<V>>(result);
+
+        return default;
+    }
+
+    /// <summary>
+    /// Removes the Component of the given type.
+    /// </summary>
+    /// <typeparam name="V">The type of the component to remove.</typeparam>
+    /// <returns>A value indicating whether the remove was successful or not.</returns>
+    public bool Remove<V>(V component) where V : T
+    {
+        if (!flatComponents.Contains(component))
+            return false;
+
+        removeValidator?.Invoke(component);
+        if (flatComponents.Remove(component))
         {
-            if (!components.TryGetValue(typeof(V), out var component))
-                return false;
+            onRemover?.Invoke(component);
 
-            removeValidator?.Invoke(component);
-            if (components.Remove(typeof(V)))
-            {
-                onRemover?.Invoke(component);
-                return true;
-            }
+            if (componentsByType.TryGetValue(typeof(V), out var components))
+                components.Remove(component);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Replaces one component with another and retrieves the replaces value when any was found
+    /// </summary>
+    /// <typeparam name="V">The type to to replace</typeparam>
+    /// <param name="toReplace">Component to search for to be replaced</param>
+    /// <param name="replacement">Component that replaces the <paramref name="toReplace"/></param>
+    /// <param name="replaced">The item that equaled <paramref name="toReplace"/> and was replaced</param>
+    /// <returns><see langword="true"/> if <paramref name="toReplace"/> was found, otherwise <see langword="false"/></returns>
```

Comment:
```suggestion
    /// Replace one component with another and retrieve the replaced value if one was found.
    /// </summary>
    /// <typeparam name="V">The type to search for and replace.</typeparam>
    /// <param name="toReplace">The component to search for to be replaced.</param>
    /// <param name="replacement">The component that replaces <paramref name="toReplace"/>.</param>
    /// <param name="replaced">The item that was equal to <paramref name="toReplace"/> and was replaced</param>
    /// <returns><see langword="true"/> if <paramref name="toReplace"/> was found, otherwise <see langword="false"/></returns>
```